src/tool_operate.c

```diff
@@ -697,45 +697,59 @@ static CURLcode post_per_transfer(struct GlobalConfig *global,
   return result;
 }
 
+/* helper function to ensure input ends in val_to_ensure */
+static void ensure_trailing(char **input, const char val_to_ensure)
+{
+  if(*input && **input) {
+    size_t len = strlen(*input);
+    if(((*input)[len - 1] != val_to_ensure)) {
+      char *composed_input = aprintf("%s%c", *input, val_to_ensure);
+      if(composed_input) {
+        Curl_safefree(*input);
+        *input = aprintf("%s", composed_input);
+        Curl_safefree(composed_input);
+      }
+    }
+  }
+}
+
 static char *ipfs_gateway(void)
 {
   char *gateway = NULL;
   char *ipfs_path = NULL;
   char *gateway_composed_file_path = NULL;
   FILE *gateway_file = NULL;
 
-  gateway = getenv("IPFS_GATEWAY");
+  gateway = curlx_getenv("IPFS_GATEWAY");
 
   /* Gateway is found from environment variable. */
   if(gateway && *gateway) {
-    char *composed_gateway = NULL;
-    bool add_slash = (gateway[strlen(gateway) - 1] != '/');
-    composed_gateway = aprintf("%s%s", gateway, (add_slash) ? "/" : "");
-    if(composed_gateway) {
-      gateway = aprintf("%s", composed_gateway);
-      Curl_safefree(composed_gateway);
-    }
+    ensure_trailing(&gateway, '/');
```

Comment:
If `gateway` is NULL when this function returns, should it still return that NULL here or continue? And if it returns NULL here, should it not provide an error message to help the user why it stops?